library/Network/Wai/Middleware/OpenTelemetry.hs

```diff
@@ -0,0 +1,62 @@
+module Network.Wai.Middleware.OpenTelemetry
+  ( newOpenTelemetryWaiMiddleware
+  , openTelemetryMiddleware
+
+  -- * Exported for testing
+  , ddTraceId
+  ) where
+
+import Freckle.App.Prelude
+
+import Blammo.Logging (withThreadContext, (.=))
+import Data.Bits (Bits(shift))
+import qualified Data.ByteString.Char8 as BS8
+import qualified Data.ByteString.Short.Internal as SBI
+import Data.Primitive.ByteArray (ByteArray(ByteArray), indexByteArray)
+import Data.Word (Word64, Word8)
+import Freckle.App.OpenTelemetry
+import Network.Wai
+import Network.Wai.Middleware.AddHeaders
+import qualified OpenTelemetry.Instrumentation.Wai as Trace
+import OpenTelemetry.Trace.Id (TraceId, traceIdBytes)
+
+newOpenTelemetryWaiMiddleware :: IO Middleware
+newOpenTelemetryWaiMiddleware = do
+  otel <- Trace.newOpenTelemetryWaiMiddleware
+  pure $ otel . openTelemetryMiddleware
+
+-- | Add 'TraceId' information to context and responses
+--
+-- - Adds @trace_id@ to the logging context
+-- - Adds @X-Datadog-TraceId@ to response headers
+--
+-- This is added automatically by our 'newOpenTelemetryWaiMiddleware'.
+--
+openTelemetryMiddleware :: Middleware
+openTelemetryMiddleware app request respond = do
+  mTraceId <- getCurrentTraceId
+  case mTraceId of
+    Nothing -> app request respond
+    Just (traceId, _) -> do
+      let
+        traceIdInt = ddTraceId traceId
+        traceIdIntBS = BS8.pack $ show traceIdInt
+
+      withThreadContext ["trace_id" .= traceIdInt] $ addHeaders
+        [("X-Datadog-Trace-Id", traceIdIntBS)]
+        app
+        request
+        respond
```

Comment:
NOTE: everything seems to function without pulling in the actual (unreleased) [Datadog propagator](https://github.com/iand675/hs-opentelemetry/blob/main/propagators/datadog/src/OpenTelemetry/Propagator/Datadog.hs), which would do similar stuff with the headers and trace-ids to make them Datadog compatible. All we're really doing here is ensuring the `trace_id` present in the logs is DD-formatted so it correlates with the traces in Datadog.